py/obj.h

```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
 // check for more specific object types.
 // Note: these are kept as macros because inline functions sometimes use much
 // more code space than the equivalent macros, depending on the compiler.
-#define mp_obj_is_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+// don't use mp_obj_is_exact_type directly; use mp_obj_is_type which provides additional safety checks.
+// use the former only if you need to bypass these checks (because you've already checked everything else)
+#define mp_obj_is_exact_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+
+// type checks are split to a separate, constant result macro. I found this way to work without hindering GCC's
+// optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type result in
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
+#define _type_checks(t) (                                                         \
+    MP_STATIC_ASSERT((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
+    MP_STATIC_ASSERT((t) != &mp_type_int), assert((t) != &mp_type_int),           \
+    MP_STATIC_ASSERT((t) != &mp_type_str), assert((t) != &mp_type_str),           \
+    MP_STATIC_ASSERT((t) != &mp_type_NoneType), assert((t) != &mp_type_NoneType), \
```

Comment:
We could add all `mp_type_fun_*` but I think that since there's no "parent type" to all of them (like there's parent type `int` for small ints and long ints), it's pretty obvious that `mp_obj_is_type(x, &mp_type_fun_bc)` checks for a bytecode function and not for a builtin function. The difference between `mp_obj_is_type` to `mp_obj_is_exact_type` is "negligible" here

Reply:
OK, agreed, don't add anything more